File: src/Classification.cc
// -*- lsst-c++ -*-
/*
 * LSST Data Management System
 * Copyright 2008-2013 LSST Corporation.
 *
 * This product includes software developed by the
 * LSST Project (http://www.lsst.org/).
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the LSST License Statement and
 * the GNU General Public License along with this program.  If not,
 * see <http://www.lsstcorp.org/LegalNotices/>.
 */
#include "ndarray/eigen.h"
#include <iostream>
#include <cmath>
#include <numeric>
#include "lsst/afw/table/Schema.h"
#include "lsst/afw/detection/Psf.h"
#include "lsst/meas/base/Classification.h"


namespace lsst { namespace meas { namespace base {


ClassificationExtras::ClassificationExtras() : probability(std::numeric_limits<double>::quiet_NaN()) {}; ///< Constructor; initializes everything to NaN

ClassificationExtrasMapper::ClassificationExtrasMapper(
        afw::table::Schema & schema,
        std::string const & prefix,
        ClassificationControl const & control
    ) : _fluxRatio(control.fluxRatio), _modelErrFactor(control.modelErrFactor), _psfErrFactor(control.psfErrFactor),
    _probability(
        schema.addField(
            afw::table::Field<double>(
                prefix + "_probability", "probability that the object is extended", ""
            ), true
        )
    )
{};

void ClassificationExtrasMapper::apply(afw::table::BaseRecord & record, ClassificationExtras const & result) const
{
    record.set(_probability, result.probability);
};

ClassificationAlgorithm::ResultMapper ClassificationAlgorithm::makeResultMapper(
    afw::table::Schema & schema, std::string const & name, Control const & ctrl)
{
    return ResultMapper(schema, name, ctrl);
}

template <typename T>
ClassificationAlgorithm::Result ClassificationAlgorithm::apply(
    afw::image::Exposure<T> const & exposure,
    float psfFlux,
    float psfFluxErr,
    float modelFlux,
    float modelFluxErr,
    Control const & ctrl
) {
    typedef typename afw::image::Exposure<T>::MaskedImageT MaskedImageT;
    typedef typename MaskedImageT::Image ImageT;
    Result result;

    // set the probability by comparing the corrected model flux to the corrected psf flux
    result.probability = (ctrl.fluxRatio*modelFlux + ctrl.modelErrFactor*modelFluxErr)
        < (psfFlux + ctrl.psfErrFactor*psfFluxErr) ? 0.0 : 1.0;

    return result;
}

ClassificationInput::Vector ClassificationInput::makeVector(afw::table::SourceCatalog const & catalog) {
    Vector r;
    r.reserve(catalog.size());
    for (afw::table::SourceCatalog::const_iterator i = catalog.begin(), end=catalog.end(); i != end; ++i) {
        r.push_back(ClassificationInput(*i));
    }
    return r;
}

ClassificationInput::ClassificationInput(afw::table::SourceRecord const & record) {
    psfFlux_ = record.getPsfFlux();
    psfFluxSigma_ = record.getPsfFluxErr();
    modelFlux_ = record.getModelFlux();
    modelFluxSigma_ = record.getModelFluxErr();
}

template <typename T>
ClassificationAlgorithm::Result ClassificationAlgorithm::apply(
    afw::image::Exposure<T> const & exposure,
    Input const & inputs,
    Control const & ctrl
) {
    return apply(exposure, inputs.psfFlux_, inputs.psfFluxSigma_, inputs.modelFlux_, inputs.modelFluxSigma_, ctrl);
}

#define INSTANTIATE(T)                                                  \
    template ClassificationAlgorithm::Result ClassificationAlgorithm::apply(          \
        afw::image::Exposure<T> const & exposure,                       \
        float psfFlux,                                                  \
        float psfFluxSigma,                                             \
        float modelFlux,                                                  \
        float modelFluxSigma,                                             \
        Control const & ctrl                                            \
    );                                                                  \
    template                                                            \
    ClassificationAlgorithm::Result ClassificationAlgorithm::apply(                   \
        afw::image::Exposure<T> const & exposure,                       \
        Input const & inputs,                                           \
        Control const & ctrl                                            \
    )

INSTANTIATE(float);
INSTANTIATE(double);

}}} // namespace lsst::meas::base

File: python/lsst/meas/base/classification.py
# This file is part of meas_base.
#
# Developed for the LSST Data Management System.
# This product includes software developed by the LSST Project
# (https://www.lsst.org).
# See the COPYRIGHT file at the top-level directory of this distribution
# for details of code ownership.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

"""Definition and registration of classification plugins.
"""

import numpy as np

import lsst.pex.config
from .catalogCalculation import CatalogCalculationPluginConfig, CatalogCalculationPlugin
from .pluginRegistry import register

__all__ = (
    "CatalogCalculationClassificationConfig", "CatalogCalculationClassificationPlugin",
)


class CatalogCalculationClassificationConfig(CatalogCalculationPluginConfig):
    """Configuration for catalog classification plugin.
    """

    fluxRatio = lsst.pex.config.Field(dtype=float, default=.95, optional=True,
                                      doc="critical ratio of model to psf flux")
    modelErrFactor = lsst.pex.config.Field(dtype=float, default=0.0, optional=True,
                                           doc="correction factor for modelFlux error")
    psfErrFactor = lsst.pex.config.Field(dtype=float, default=0.0, optional=True,
                                         doc="correction factor for psfFlux error")


@register("base_ClassificationExtendedness")
class CatalogCalculationClassificationPlugin(CatalogCalculationPlugin):
    """Plugin which calculates a binary measure of source extendedness.

    Extendedness is based on a simple cut of the ratio of the PSF flux to the
    model flux.

    Notes
    -----
    Because the fluxes on which this algorithm is based on are slot
    measurements, they can be provided by different algorithms, and the
    `~CatalogCalculationClassificationConfig.fluxRatio` threshold used by this
    algorithm should generally be set differently for different algorithms.
    To do this, plot the difference between the PSF magnitude and the model
    magnitude vs. the PSF magnitude, and look for where the cloud of galaxies
    begins.
    """

    ConfigClass = CatalogCalculationClassificationConfig

    @classmethod
    def getExecutionOrder(cls):
        return cls.DEFAULT_CATALOGCALCULATION

    def __init__(self, config, name, schema, metadata):
        CatalogCalculationPlugin.__init__(self, config, name, schema, metadata)
        self.keyProbability = schema.addField(name + "_value", type="D",
                                              doc="Set to 1 for extended sources, 0 for point sources.")
        self.keyFlag = schema.addField(name + "_flag", type="Flag", doc="Set to 1 for any fatal failure.")

    def calculate(self, measRecord):
        modelFlux = measRecord.getModelInstFlux()
        psfFlux = measRecord.getPsfInstFlux()
        modelFluxFlag = (measRecord.getModelFluxFlag()
                         if measRecord.table.getModelFluxSlot().isValid()
                         else False)
        psfFluxFlag = (measRecord.getPsfFluxFlag()
                       if measRecord.table.getPsfFluxSlot().isValid()
                       else False)
        flux1 = self.config.fluxRatio*modelFlux
        if self.config.modelErrFactor != 0:
            flux1 += self.config.modelErrFactor*measRecord.getModelInstFluxErr()
        flux2 = psfFlux
        if not self.config.psfErrFactor == 0:
            flux2 += self.config.psfErrFactor*measRecord.getPsfInstFluxErr()

        # A generic failure occurs when either FluxFlag is set to True
        # A generic failure also occurs if either calculated flux value is NaN:
        #     this can occur if the Flux field itself is NaN,
        #     or the ErrFactor != 0 and the FluxErr is NaN
        if np.isnan(flux1) or np.isnan(flux2) or modelFluxFlag or psfFluxFlag:
            self.fail(measRecord)
        else:
            measRecord.set(self.keyProbability, 0.0 if flux1 < flux2 else 1.0)

    def fail(self, measRecord, error=None):
        measRecord.set(self.keyFlag, True)
